java-squid/src/main/java/org/sonar/java/resolve/TypeResolutionHelper.java

```diff
@@ -0,0 +1,220 @@
+   */
+  private static Multimap<JavaType, ParametrizedTypeJavaType> relevant(Set<JavaType> mec, Multimap<JavaType, JavaType> st) {
+    Multimap<JavaType, ParametrizedTypeJavaType> relevants = HashMultimap.create();
+    for (JavaType g : mec) {
+      ImmutableSet.Builder<ParametrizedTypeJavaType> results = ImmutableSet.builder();
+      String fullyQualifiedName = g.fullyQualifiedName();
+      for (Collection<JavaType> supertypes : st.asMap().values()) {
+        results.addAll(relevant(fullyQualifiedName, supertypes));
+      }
+      relevants.putAll(g, results.build());
+    }
+    return relevants;
+  }
+
+  private static Set<ParametrizedTypeJavaType> relevant(String fullyQualifiedName, Collection<JavaType> supertypes) {
+    ImmutableSet.Builder<ParametrizedTypeJavaType> results = ImmutableSet.builder();
+    for (JavaType v : supertypes) {
+      if (v instanceof ParametrizedTypeJavaType && ((ParametrizedTypeJavaType) v).rawType.is(fullyQualifiedName)) {
+        results.add((ParametrizedTypeJavaType) v);
+      }
+    }
+    return results.build();
+  }
+
+  /**
+   * Let LCP, the "least containing parameterization" be the most specific parameterization of
+   * the generic type G that contains all the relevant parameterizations of G.
+   * @param relevants
+   * @return
+   */
+  private static Map<JavaType, ParametrizedTypeJavaType> leastContainingParameterization(Multimap<JavaType, ParametrizedTypeJavaType> relevants) {
+    Map<JavaType, ParametrizedTypeJavaType> results = Maps.newHashMap();
+    for (Entry<JavaType, Collection<ParametrizedTypeJavaType>> g : relevants.asMap().entrySet()) {
+      results.put(g.getKey(), lcp(Lists.newArrayList(g.getValue())));
+    }
+    return results;
+  }
+
+  private static ParametrizedTypeJavaType lcp(List<ParametrizedTypeJavaType> types) {
+    return lcp(types.get(0), types.subList(1, types.size()));
+  }
+
+  private static ParametrizedTypeJavaType lcp(ParametrizedTypeJavaType lcpCandidate, List<ParametrizedTypeJavaType> rest) {
+    if (rest.isEmpty()) {
+      return lcpCandidate;
+    }
+    return lcp(lcp(lcpCandidate, rest.get(0)), rest.subList(1, rest.size()));
+  }
+
+  private static ParametrizedTypeJavaType lcp(ParametrizedTypeJavaType t1, ParametrizedTypeJavaType t2) {
+    TypeSubstitution substitution = new TypeSubstitution();
+    for (TypeVariableJavaType typeVariable : t1.typeParameters()) {
+      JavaType t1Substitution = t1.substitution(typeVariable);
+      JavaType t2Substitution = t2.substitution(typeVariable);
+      substitution.add(typeVariable, leastContainingTypeArgument(t1Substitution, t2Substitution));
+    }
+    return new ParametrizedTypeJavaType(t1.rawType.symbol, substitution);
+  }
+
+  private static JavaType leastContainingTypeArgument(JavaType t1, JavaType t2) {
+    if (isInvalid(t1) || isInvalid(t2)) {
+      return Symbols.unknownType;
+    } else if (t1.equals(t2)) {
+      return t1;
+    }
+    // FIXME types of wildcard trees are not implemented
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Take the required action to fix the issue indicated by this comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1134)